### ISSUE TITLE
User-controlled target processor kind

### ISSUE THREAD

**Issue by manopapad** [nv-legate/legate.core#422]
@rohany brought up a scenario where the user would want to constrain one phase of the application to run on CPUs, and then switch to GPUs.

My first hunch would be to allow the user to annotate parts of the code with a processor target, e.g. using a python context:

```
with legate.execute_on_cpus():
  x = make_input()
with legate.execute_on_gpus():
  y = process(x)
```

But maybe Rohan can explain his usecase a bit more, and we can see if there's a better way to support it.

**Comment by rohany** [comment]
As manolis sketched, I have an application which has a data generation phase that is very memory hungry (especially for deferred allocations) and then a phase that has pretty stable memory usage (a solve). I would like to perform the data generation (and allocate more deferred memory) on the CPUs which have more memory, and then flip a switch to have the solve run on the GPUs. This seems like something that could be controlled by legate core, as individual libraries can control this pretty well but cannot control what other legate libraries will do.

**Comment by magnatelee** [comment]
for everyone's benefit, here's a snippet that shows a tentative API design:
<img width="1354" alt="all devices = get machine()" src="https://user-images.githubusercontent.com/331175/197047561-926ba9f7-ef0e-47df-8fd4-3222a2912b7e.png">

though the API looks simple, the resource scoping has major consequences on how the core internally manages cached states (as #439 already explored some of it). To properly address them, a new scoping mechanism would be introduced to centralize all cached states sensitive to the resource configuration, such as the max degree of parallelism, key partitions, communicators, etc.

**Comment by RAMitchell** [comment]
My 2 cents. Out of memory errors are our most significant barrier to running distributed XGBoost. We can tell users to launch more machines but we do want the utilisation of each machine to be high and this is more often limited by memory.

Allowing user-controlled target processors would be nice, but it would be even nicer if legate could automatically recognise that a task will exceed available GPU memory and run it on a CPU if there is more memory available. Maybe this is asking too much given that things like unbound store sizes are not known ahead of time?

What if I were able to label a task somehow as "memory bound", indicating that legate should do everything possible to avoid OOM.

**Comment by magnatelee** [comment]
> Allowing user-controlled target processors would be nice, but it would be even nicer if legate could automatically recognise that a task will exceed available GPU memory and run it on a CPU if there is more memory available. Maybe this is asking too much given that things like unbound store sizes are not known ahead of time?

Right, there are some uncertainties as to how much memory a given task will consume on unbound stores and also temporary allocations that do not escape the task. so expecting a runtime system to fully automate this is both too much to ask and dangerous, as such a heuristic is often obscure and hard (or even impossible) to tune. User-controlled resource scoping is  actually designed to provide the resource decomposition and containment, though it happens to be helpful in @rohany's case of the input data not fitting to the GPU memory. 

> What if I were able to label a task somehow as "memory bound", indicating that legate should do everything possible to avoid OOM.

My plan for memory oversubscription cases (or out-of-core processing as Mike would put it) requires more than simply labelling individual tasks; what we need is a workflow description, which really is just a Python function with some decorator invoking a sequence of Legate library API calls. Two necessary operational properties for the oversubscription are that the size of the output for this workflow is strictly smaller than the inputs', and that there must exist a (horizontal) decomposition of the workflow such that each slice of the decomposition would fit to the memory. This needs some amount of JIT analysis on a workflow, but is doable within Legate. I'm actually supposed to draft a design doc on this, so hopefully I'll send it out once it's ready.
